File: cpp/53.cpp
#include <iostream>
#include <vector>
#include<algorithm>
using namespace std;

class Solution 
{
public:
    int maxSubArray(vector<int>& nums) 
    {
        if(nums.size()==0)
            return 0;
        int res=INT_MIN;
        int fn=-1;
        for (int i = 0; i < nums.size(); i++)
        {
            fn=max(fn+nums[i],nums[i]);
            res=max(fn,res);
        }
        return res;
    }
};

File: java/53.java
class Solution {
    public int maxSubArray(int[] nums) {
        if(nums.length==1)
            return nums[0];

        int[] res = new int[nums.length];

        res[0] = nums[0];

        int maxn = res[0];

        for(int i=1; i< nums.length; i++){
            res[i] = Math.max(res[i-1]+nums[i], nums[i]);
            if(res[i] > maxn)
                maxn = res[i];
        }


        return maxn;
    }
}